Sorts Prow configs by test job names

yguo0905 opened the pull request:
Extracted the changes for sorting configs in https://github.com/kubernetes/test-infra/pull/3115 into this separate PR.

This PR make future config changes (including the configs generated automatically) easier to review.

/assign @krzyzacy 
/assign @spxtr

krzyzacy commented:
how about combine with jobs/config_sort.py?

yguo0905 commented:
PTAL. What should be done to fix the pull-test-infra-bazel error?

krzyzacy commented:
I'm not familiar with this workspace set up either, @spxtr @ixdy might be giving better answer here

ixdy commented on jobs/BUILD:
```diff
@@ -10,6 +10,13 @@ filegroup(
 )
 
 filegroup(
+    name = "scripts",
```
rather than expose config_sort through the filegroup, you should export it as a `py_binary` or `py_library` which depends on the `ruamel_yaml` dependency you added, and then depend on that new rule in `prow/config/BUILD`.

yguo0905 replied:
Thanks. I will try.

ixdy commented on prow/config/config_test.go:
```diff
@@ -294,3 +297,17 @@ func TestBazelJobHasContainerPort(t *testing.T) {
+	if err != nil {
+		t.Fatalf("Could not load config: %v", err)
+	}
+	expected, err := exec.Command("../../jobs/config_sort.py", "--dump-sorted-prow-config").CombinedOutput()
```
maybe import `jobs/config_sort` and call a function rather than shelling out, since this is all python?

yguo0905 replied:
We are invoking the `jobs/config_sort.py` from the Go test, not Python?

ixdy replied:
you are right. I somehow missed that this was Go, not Python.

yguo0905 commented:
I'm not sure how to resolve this error. Do you have any ideas?

In `jobs/config_sort.py`, I have:
```
import ruamel_yaml as yaml    # ruamel_yaml is the workspace name
```

`bazel run //jobs:config_sort` gave me this.

```
INFO: Found 1 target...
Target //jobs:config_sort up-to-date:
  bazel-bin/jobs/config_sort
INFO: Elapsed time: 0.453s, Critical Path: 0.00s

INFO: Running command line: bazel-bin/jobs/config_sort
Traceback (most recent call last):
  File "/usr/local/google/home/ygg/.cache/bazel/_bazel_ygg/d577eee115221f99cc98b3863293dcdd/execroot/test-infra/bazel-out/local-fastbuild/bin/jobs/config_sort.runfiles/__main__/jobs/config_sort.py", line 24, in <module>
    import ruamel_yaml as yaml
  File "/usr/local/google/home/ygg/.cache/bazel/_bazel_ygg/d577eee115221f99cc98b3863293dcdd/execroot/test-infra/bazel-out/local-fastbuild/bin/jobs/config_sort.runfiles/ruamel_yaml/__init__.py", line 69, in <module>
    from ruamel.yaml.main import *  # NOQA
ImportError: No module named ruamel.yaml.main
ERROR: Non-zero return code '1' from command: Process exited with status 1.
```

ixdy commented:
The issue seems to be that normally `ruamel.yaml` is installed in `$PYTHONPATH/ruamel/yaml`, but here it's in `ruamel_yaml`, hence python is having trouble resolving it.

I'm not sure how we can coax bazel into getting the external dependency installed in `ruamel/yaml`. @spxtr any ideas?

spxtr commented:
Not sure. Maybe ask bazel-discuss or StackOverflow?

yguo0905 commented:
/test pull-test-infra-bazel

yguo0905 commented:
PTAL.

- Added a custom bazel rule `http_archive_with_pkg_path` in `test_infra.bzl` (let me know if there's better name) as suggested [here](https://groups.google.com/d/msg/bazel-discuss/3CWlklvjxWg/I9a5JtyuAAAJ).
- Moved the test (checking whether the Prow configs are sorted in presubmits) from `prow/config_test.go` to `bootstrap_test.py` because 1) we moved the sorting logic from `prow/sort_config.py` to `jobs/sort_config.py` and 2) it seems impossible to add a python dependency (`jobs/sort_config.py` in a go_test (`prow/config_test.go`).

krzyzacy commented:
lgtm, that's good to learn! Let bazel experts @ixdy and @spxtr also take a look and make the final stamp.

ixdy commented on jenkins/bootstrap_test.py:
```diff
@@ -2285,7 +2286,16 @@ def test_config_is_sorted(self):
+                          '`bazel run //jobs:config_sort`')
+        with open(bootstrap.test_infra('prow/config.yaml')) as fp:
+            original = fp.read()
+            sys.path.insert(1, os.path.join(sys.path[0], '../jobs'))
```
this is kinda gross.
can we just make a new test in `jobs/` instead?

maybe move this and the config.json test out of bootstrap_test.py, since neither belongs here.

krzyzacy replied:
yeah I'd like to move a whole bunch of things from `bootstrap_test.py` to `jobs/`

yguo0905 replied:
Moved to `jobs/config_test.py`. PTAL.

ixdy commented on jobs/config_sort.py:
```diff
@@ -34,5 +50,32 @@ def sort():
         fp.truncate()
 
+
+def sorted_prow_config():
```
this logic should arguably live in `prow/`, not `jobs/`.

ixdy replied:
hm, I see that @krzyzacy originally recommended combining these. OK.

spxtr replied:
I don't think it should live in `prow/`, since prow doesn't care if they're sorted or not.

yguo0905 commented:
Just rebased. `prow/config.yaml` changes frequently. Please take a look.

krzyzacy commented:
@ixdy does it look good from the bazel side?

yguo0905 commented:
Friendly ping. @ixdy

spxtr commented on test_infra.bzl:
```diff
@@ -0,0 +1,19 @@
+        output=ctx.attr.pkg_path)
+    ctx.file(ctx.attr.pkg_path+"/BUILD.bazel", ctx.attr.build_file_content)
+
+http_archive_with_pkg_path = repository_rule(
```
Please add a comment describing this rule.

yguo0905 commented:
Done. Thanks.

@krzyzacy, I think this is ready to merge unless you have more comments.

krzyzacy commented:
/lgtm

ixdy commented:
(delayed lgtm, thanks for all of the fixes!)
